### PR TITLE
RPC refactoring: Access wallet using new GetWalletForJSONRPCRequest

### PR DESCRIPTION
**Pull request by luke-jr**
Part of the refactorings needed for basic multiwallet (#8694)

### CODE REVIEW

**Comment by jonasschnelli** [comment]
I don't like the coupling and the `#ifdef ENABLE_WALLET` in `rpc/server.cpp|.h`.
I'd recommend to keep the `CRPCRequestInfo` wallet-free.

I think we should pack the request path (URI) into the `CRPCRequestInfo` and or informations about the authentication (in case we want to distinct wallets based on authentication).
Then I think there should be a method in `wallet.cpp` (or in `rpcwallet.cpp`) that maps a `CWallet *` pointer from a given URI, Auth-Info or the complete `CRPCRequestInfo` instance.

Instead of the `CWallet *& pwallet = reqinfo.wallet;` there could be then something like `CWallet *pwallet = CWallets::getWalletFromRequest(reqinfo)`

**Comment by jonasschnelli** [comment]
General ConceptACK on a `CRPCRequestInfo` for the RPC table commands.
Maybe it could also include the `UniValue params` and `fHelp`?

File: src/rpc/misc.cpp [hunk 1]
```diff
@@ -70,7 +70,8 @@ UniValue getinfo(const UniValue& params, bool fHelp)
         );
 
 #ifdef ENABLE_WALLET
-    LOCK2(cs_main, pwalletMain ? &pwalletMain->cs_wallet : NULL);
+    CWallet *& pwallet = reqinfo.wallet;
```
Comment (paveljanik) [inline src/rpc/misc.cpp]: `*&`?

Comment (luke-jr) [inline src/rpc/misc.cpp]: Yes, pwallet is an alias to reqinfo.wallet which is of type CWallet*.

**Comment by luke-jr** [comment]
> I think we should pack the request path (URI) into the CRPCRequestInfo and or informations about the authentication (in case we want to distinct wallets based on authentication).
Then I think there should be a method in wallet.cpp (or in rpcwallet.cpp) that maps a CWallet * pointer from a given URI, Auth-Info or the complete CRPCRequestInfo instance.

That sounds nice, but greatly complicates the implementation. :(

**Comment by MarcoFalke** [comment]
I think this can be closed after #8788?

**Comment by jonasschnelli** [comment]
Closing in favor of merged #8788

**Comment by luke-jr** [comment]
Rebased and refactored based on @jonasschnelli 's idea.

**Comment by laanwj** [comment]
Makes sense, utACK ab5ce98

**Comment by luke-jr** [comment]
Rebased and addressed nit

**Comment by instagibbs** [comment]
utACK 7de55733c5690350c48ec94660b5be56fbb5eb39

**Comment by luke-jr** [comment]
Minor change: Forward-declared CWallet even for non-wallet builds so it can be used in a pointer type, avoiding unnecessary casting.

File: src/rpc/rawtransaction.cpp [hunk 2]
```diff
@@ -35,6 +35,11 @@
 
 using namespace std;
 
+#ifdef ENABLE_WALLET
+std::string HelpRequiringPassphrase(CWallet *);
```
Comment (TheBlueMatt) [inline src/rpc/rawtransaction.cpp]: I'd kinda prefer we add another declaration outside of a header in an unrelated file...can we keep it in server.h and ifdef ENABLE_WALLET it?

Comment (luke-jr) [inline src/rpc/rawtransaction.cpp]: I suppose we could. But then it's more likely to get used in new code (which I think we want to discourage?)

Comment (TheBlueMatt) [inline src/rpc/rawtransaction.cpp]: I'd think we can (and, more importantly, would) still nag people submitting PRs which add more uses of it outside of src/wallet/rpc*.

File: src/rpc/misc.cpp [hunk 3]
```diff
@@ -26,6 +26,10 @@
 
 using namespace std;
 
+#ifdef ENABLE_WALLET
+CWallet *GetWalletForJSONRPCRequest(const JSONRPCRequest&);
```
Comment (TheBlueMatt) [inline src/rpc/misc.cpp]: Here as well.

File: src/wallet/rpcdump.cpp [hunk 4]
```diff
@@ -29,6 +29,7 @@
 
 using namespace std;
 
+CWallet *GetWalletForJSONRPCRequest(const JSONRPCRequest&);
 void EnsureWalletIsUnlocked(CWallet *);
 bool EnsureWalletIsAvailable(CWallet *, bool avoidException);
```
Comment (TheBlueMatt) [inline src/wallet/rpcdump.cpp]: Can we move these to some rpc.h in src/wallet? Why are we declaring functions in another file from the definition if they're both in src/wallet???

File: src/wallet/rpcwallet.cpp [hunk 5]
```diff
@@ -30,6 +30,11 @@ using namespace std;
 int64_t nWalletUnlockTime;
 static CCriticalSection cs_nWalletUnlockTime;
 
+CWallet *GetWalletForJSONRPCRequest(const JSONRPCRequest& request)
```
Comment (TheBlueMatt) [inline src/wallet/rpcwallet.cpp]: Can you document this function somewhere (also probably move its definition to src/wallet/rpcwallet.h) - a few places in src/wallet assume it always returns non-NULL but this is not documented (I havent looked at the actual multi-wallet PR, but does this function then throw an RPC exception if you ask for a wallet that isnt loaded, or does it return NULL then)?

**Comment by gmaxwell** [comment]
Concepts ACK.

So in some cases in the RPC you get the wallet pointer from json but then the check if it's available is far below.  This is begging for someone to insert code that uses a potentially null pointer between to two and doesn't notice because their function doesn't get tested with a missing wallet.  I would recommend moving the creation of that local pointer down to the point where you're going to test it.

Alternatively or in addition,  rename GetWalletForJSONRPCRequest to TryGetWalletForJSONRPCRequest and make GetWalletForJSONRPCRequest a wrapper for it that throws when it fails?

**Comment by luke-jr** [comment]
I liked the TryGetWalletForJSONRPCRequest/GetWalletForJSONRPCRequest refactor idea, but I don't see any clean way to do it without changing the `help` behaviours.

So I moved the one case EnsureWalletIsAvailable was delayed up, and removed the blank line between them and GetWalletForJSONRPCRequest.

File: src/wallet/rpcwallet.cpp [hunk 6]
```diff
@@ -2019,9 +2016,8 @@ UniValue walletpassphrase(const JSONRPCRequest& request)
     pwallet->TopUpKeyPool();
 
     int64_t nSleepTime = request.params[1].get_int64();
-    LOCK(cs_nWalletUnlockTime);
-    nWalletUnlockTime = GetTime() + nSleepTime;
-    RPCRunLater("lockwallet", boost::bind(LockWallet, pwallet), nSleepTime);
+    pwallet->nRelockTime = GetTime() + nSleepTime;
+    RPCRunLater(strprintf("lockwallet_%u", uintptr_t(pwallet)), boost::bind(LockWallet, pwallet), nSleepTime);
```
Comment (TheBlueMatt) [inline src/wallet/rpcwallet.cpp]: I super dont like the fact that the pointer to the wallet will end up in debug.log if you have debug=rpc enabled.

File: src/wallet/rpcwallet.cpp [hunk 7]
```diff
@@ -340,8 +349,7 @@ UniValue getaddressesbyaccount(const JSONRPCRequest& request)
 
     // Find all addresses that have the given account
     UniValue ret(UniValue::VARR);
-    BOOST_FOREACH(const PAIRTYPE(CBitcoinAddress, CAddressBookData)& item, pwallet->mapAddressBook)
-    {
+    for (const PAIRTYPE(CBitcoinAddress, CAddressBookData)& item : pwallet->mapAddressBook) {
```
Comment (TheBlueMatt) [inline src/wallet/rpcwallet.cpp]: nit: I believe you can remove the PAIRTYPE and replace with std::pair since we're not using BOOST_FOREACH now.

File: src/wallet/rpcwallet.cpp [hunk 8]
```diff
@@ -1219,8 +1241,7 @@ UniValue ListReceived(CWallet * const pwallet, const UniValue& params, bool fByA
     // Reply
     UniValue ret(UniValue::VARR);
     map<string, tallyitem> mapAccountTally;
-    BOOST_FOREACH(const PAIRTYPE(CBitcoinAddress, CAddressBookData)& item, pwallet->mapAddressBook)
-    {
+    for (const PAIRTYPE(CBitcoinAddress, CAddressBookData)& item : pwallet->mapAddressBook) {
```
Comment (TheBlueMatt) [inline src/wallet/rpcwallet.cpp]: nit: Here as well (no PAIRTYPE).

File: src/wallet/rpcwallet.cpp [hunk 9]
```diff
@@ -1635,14 +1665,14 @@ UniValue listaccounts(const JSONRPCRequest& request)
             includeWatchonly = includeWatchonly | ISMINE_WATCH_ONLY;
 
     map<string, CAmount> mapAccountBalances;
-    BOOST_FOREACH(const PAIRTYPE(CTxDestination, CAddressBookData)& entry, pwallet->mapAddressBook) {
-        if (IsMine(*pwallet, entry.first) & includeWatchonly) // This address belongs to me
+    for (const PAIRTYPE(CTxDestination, CAddressBookData)& entry : pwallet->mapAddressBook) {
```
Comment (TheBlueMatt) [inline src/wallet/rpcwallet.cpp]: nit: Here as well (no PAIRTYPE).

**Review by TheBlueMatt — COMMENTED** [review]
Please do not tag the 4th commit "Trivial". We usually define trivial as doesnt touch any code.

As for the 6th commit: please do not do this. nothing in src/wallet is built when ENABLE_WALLET is off, so generally src/wallet/* should never be included outside of src/wallet/*

Aside from the (partial-revert) of "Move wallet RPC declarations to rpcwallet.h" and the printing of the raw pointers to debug log, this looks good to me at d9aff6e322fd4cdba8d338d304a627af46feb7eb.

**Comment by gmaxwell** [comment]
@luke-jr Looks good to me, will test as soon as you update for matt's latest comments.

**Comment by luke-jr** [comment]
Made changes requested by @TheBlueMatt, and rebased to resolve conflict. Also includes a quick sanity check that -wallet doesn't include path separators.

**Comment by TheBlueMatt** [comment]
While you're at it, can you call SanitizeString (in addition to the "/\" check) on the wallet param?

**Comment by TheBlueMatt** [comment]
utACK 7d1228bf7f1c398d9fcda6161e6aebfc5e5e3700

**Comment by gmaxwell** [comment]
Needs rebase.

**Comment by gmaxwell** [comment]
ACK.

**Comment by fanquake** [comment]
Needs rebasing again.

**Comment by luke-jr** [comment]
Rebased yet again.

**Comment by gmaxwell** [comment]
still ACK

File: src/wallet/wallet.cpp [hunk 10]
```diff
@@ -3740,6 +3740,12 @@ bool CWallet::InitLoadWallet()
 
     std::string walletFile = GetArg("-wallet", DEFAULT_WALLET_DAT);
```
Comment (BlockMechanic) [inline src/wallet/wallet.cpp]: Should this not be "count" i.e for multiple wallet files to be created ? Adding perhaps a for loop to iterate over required num of wallets?

Comment (luke-jr) [inline src/wallet/wallet.cpp]: No, because this PR is not multiwallet...

File: src/rpc/misc.cpp [hunk 11]
```diff
@@ -234,6 +235,9 @@ UniValue validateaddress(const JSONRPCRequest& request)
     return ret;
 }
 
+// Needed even with !ENABLE_WALLET, to pass (ignored) pointers around
+class CWallet;
```
Comment (ryanofsky) [inline src/rpc/misc.cpp]: This doesn't actually seem to be needed since CWallet is also forward declared in init.h, but maybe it is better to keep it.

**Review by ryanofsky — COMMENTED** [review]
utACK d678771c660c936a6222dd43a204cb2699100f3e

**Comment by jonasschnelli** [comment]
Great PR. I think we should get this in as soon as possible.
utACK d678771c660c936a6222dd43a204cb2699100f3e

**Comment by laanwj** [comment]
utACK d678771

**Comment by jtimon** [comment]
By the way, I reviewed this partially after merged when rebasing #9845 (making it mostly pointless since most was done here already).

**Comment by adneerav** [comment]
Hello @luke-jr ,

Is there any way where we can increase the number of wallets to be in loaded state at same time.As rightnow its like if I load wallet more than 100 or 150 daemon is getting crashed/stopped. I am trying to load/unload wallet dynamically using RPC.

What i tried is incraesing the `set_lk_max_lockers` to 400000 and also added
`set_lg_regionmax 1048576` But still same things I am facing.

Reflink for above setup : https://docs.oracle.com/cd/E17276_01/html/api_reference/C/set_lk_max_lockers_parameter.html
https://bitcoin.stackexchange.com/a/70737

So is there any way or configuration or any changes which can increase and allow more number of wallets to be loaded ?
